File: Extras/servers/python/launch.py
#!/usr/bin/env python3
'''
Simple take-off-and-move-forward script

Copyright (C) 2021 Simon D. Levy

MIT License
'''

import cv2
import numpy as np
import argparse
from argparse import ArgumentDefaultsHelpFormatter

from launch_controller import LaunchController
from multicopter_server import MulticopterServer
from mixers import PhantomMixer, IngenuityMixer


class LaunchCopter(MulticopterServer):

    def __init__(
            self,
            mixer,
            kp=1.0,
            ki=0.0,
            initial_target=15.0):

        MulticopterServer.__init__(self)

        self.mixer = mixer

        self.time = 0
        self.target = initial_target

        # Create PID controller
        self.ctrl = LaunchController(kp, ki)

    def handleImage(self, image):
        gray = cv2.cvtColor(image, cv2.COLOR_BGR2GRAY)
        edges = cv2.Canny(gray, 100, 200)
        cv2.imshow('Edge Detection', edges)
        cv2.waitKey(1)

        nonzero = np.nonzero(edges)[0]

        # Ignore image for first five seconds
        if len(nonzero) > 0 and np.mean(nonzero) > 390 and self.time > 5:
            self.target = 30

    def getMotors(self, t, state, _stickDemands):

        # Track current time to share it with handleImage()
        self.time = t

        # Extract altitude from state.  Altitude is in NED coordinates, so we
        # negate it to use as input to PID controller.
        z = -state[MulticopterServer.STATE_Z]
        dzdt = -state[MulticopterServer.STATE_DZ]

        # Get demands U [throttle, roll, pitch, yaw] from PID controller,
        # ignoring stick demands
        u = self.ctrl.getDemands(self.target, z, dzdt, t)

        # Use mixer to convert demands U into motor values Omega
        omega = self.mixer.getMotors(u)

        # Constrain motor values to [0,1]
        omega[omega > 1] = 1
        omega[omega < 0] = 0

        # Return motor values
        return omega


def main():

    parser = argparse.ArgumentParser(
            formatter_class=ArgumentDefaultsHelpFormatter)

    parser.add_argument('--vehicle', required=False, default='Phantom',
                        help='Vehicle name')

    args = parser.parse_args()

    d = {'Phantom': PhantomMixer, 'Ingenuity': IngenuityMixer}

    if args.vehicle in d:
        copter = LaunchCopter(d[args.vehicle]())
        copter.start()

    else:
        print('Unrecognized vehicle: %s' % args.vehicle)


main()

File: Extras/attic/servers/java/Launch.java
/*
   Java Multicopter test 

   Copyright(C) 2019 Simon D.Levy

   MIT License
 */

public class Launch {

    // Target params
    static final double ALTITUDE_TARGET = 10;

    // PID params
    static final double ALT_P = 1.0;
    static final double VEL_P = 1.0;
    static final double VEL_I = 0;

    public static void main(String [] args)
    {
        // Create PID controller
        LaunchPidController pid  = new LaunchPidController(ALTITUDE_TARGET, ALT_P, VEL_P, VEL_I);

        // Create mixer
        Mixer mixer = new Mixer();

        // Create a multicopter simulation
        Multicopter copter = new Multicopter();

        // Start the simulation
        copter.start();

        // Loop until user hits the stop button
        while (true) {

            // Get time from sim
            double t = copter.getTime();

            // Negative time means quit
            if (t < 0) break;

            // Get state from sim
            double [] state = copter.getVehicleState();

            // Extract altitude and its first derivative from state.  Launch
            // is in NED coordinates, so we negate it to use as input to PID
            // controller.
            double z = -state[Multicopter.STATE_Z];
            double dz = -state[Multicopter.STATE_DZ];

            // Start with zeros for demands
            double []  omega = new double[4];


            // Get demands from PID controller
            double [] u = pid.getDemands(z, dz, t);

            // Run demands through mixer to get motor spins
            omega = mixer.getMotors(u);

            // Constrain correction to [0,1] to represent motor value
            for (int k=0; k<4; ++k) {
                omega[k] = Math.max(0., Math.min(1., omega[k]));
            }

            // Set motor values in sim
            copter.setMotors(omega);

            // Yield to Multicopter thread
            try {
                Thread.sleep(1);
            }
            catch (Exception e) {
            }

        }
    }

    // Helper function like Matlab/Numpy ones()
    static double [] ones(double x, int n)
    {
        double []  v = new double [n];
        for (int i=0; i<n; ++i) {
            v[i] = x;
        }
        return v;
    }
}
